osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs

```diff
@@ -0,0 +1,179 @@
+        }
+
+        /// <summary>
+        /// Removes an instant of spin.
+        /// </summary>
+        /// <param name="currentTime">The current time.</param>
+        /// <param name="delta">The rate-independent, instantaneous delta of the angle moved through. Negative values represent counter-clockwise movements, positive values represent clockwise movements.</param>
+        public void RemoveDelta(double currentTime, float delta)
+        {
+            while (currentTime < currentTurn.StartTime)
+            {
+                // When crossing over a turn, we need to adjust the delta so that it's relative to the end point of the next turn.
+                //
+                // This is done by ADDING the delta between the current turn and the next turn.
+                // To understand why this is, notice that delta is a rate-independent value. Suppose the turn values are { 90, 45 } (i.e. CW then CCW spin)...
+                // - If delta < 0 (e.g. -15) (i.e. CCW rotation), then the next turn should be <45 (therefore delta = -15 + (45 - 90) = -60, next = 30).
+                // - If delta = 0, then the next turn should be =45 (therefore delta = 0 + (45 - 90) = -45, next = 45).
+                // - If delta > 0 (e.g. +15) (i.e. CW rotation), then the next turn should be >45 (therefore delta = 15 + (45 - 90) = -30, next = 60).
+                //
+                // There is a special case when crossing a complete spin, because the turn following it starts at 0 rather than the previous turn's value.
+                // In this case, only the remaining delta in the current turn needs to be considered.
+
+                Turn nextTurn = turningPoints.Pop();
+
+                if (nextTurn.IsCompleteSpin)
+                    delta += currentTurn.Angle;
+                else
+                    delta += currentTurn.Angle - nextTurn.Angle;
+
+                currentTurn = nextTurn;
+            }
+
+            currentTurn.Angle += delta;
+
+            // Note: Enumerating through a stack is already reverse order.
+            currentMaxRotation = turningPoints.Prepend(currentTurn).TakeWhile(t => !t.IsCompleteSpin).Select(t => Math.Abs(t.Angle)).Max();
+            TotalRotation = 360 * turningPoints.Count(t => t.IsCompleteSpin) + currentMaxRotation;
+        }
+
+        /// <summary>
+        /// Finishes the current turn and starts a new one from its end point.
+        /// </summary>
+        /// <param name="currentTime">The start time of the new turn.</param>
+        /// <param name="direction">The turning direction.</param>
+        private void beginNewTurn(double currentTime, int direction)
+        {
+            turningPoints.Push(currentTurn);
+            currentTurn = new Turn(currentTime, direction) { Angle = currentTurn.Angle };
+        }
+
+        /// <summary>
+        /// Represents a single direction turn of the spinner.
+        /// </summary>
+        private struct Turn
```

Comment:
Can we perhaps rename this to ~~"Moment"~~ "TurningPoint" instead of "Turn"?

"One turn" to most people would describe a full rotation, which isn't the case here, so it's quite confusing.

Reply:
Before I do this, can you please define _all_ terminology here? Because we have different ways of thinking about things - I think of a turn as a varying amount, a spin as a full rotation, and a rotation as an aggregate of all spins. I've already taken a different path from that in this PR which is to think of a "spin" as a "rotation" which my mind already struggles with.

Please define:
- A word for what is basically a arbitrary angle, or what is currently `Turn`.
- A word for the "arbitrary angle" as above, that is currently being updated, or what is currently `currentTurn`.
- A word for the maximum absolute angle in any one direction starting from 0 for the current 360-degree... rotation, or what is currently `currentMaxRotation`.
- A word for the total aggregate rotation - the summation of all full 360-degree spins plus the current partial one (`currentMaxRotation`), or what is currently `TotalRotation`.

Reply:
For example, would you feel comfortable with:
```
private struct TurningPoint { ... }

private Stack<TurningPoint> TurningPoints = ...

private TurningPoint currentTurningPoint;

public void AddDelta(...)
{
    currentTurningPoint.Angle = ...
}
```

Reply:
I've applied naming changes as discussed on discord, will resolve this for now as it's feeling good to me.